### PR TITLE
Compile released code to `ESNext`; browser bundle to `ES2017`

### LINKED ISSUE #158: Target `esnext` when building releases
Consumers of this package build for a variety of targets, from the most modern to the most deprecated. The choice of which target to build their application against rests with them, and them alone.

This library builds releases against a target too.

https://github.com/elpheria/rpc-websockets/blob/4cbd246c125c7b8d6c9eada39a8875a93991a914/tsconfig.json#L7

The reason this is bad is because this process is _lossy_. If a consumer of this library wants to build for `es2020`, it's too late – you've already _downcompiled_ `rpc-websockets` for `es2017`.

This steals away from your consumers bundle size optimizations, compatibility, and sometimes even performance. Consider this consumer for instance, whose customers can no longer use the software on the Vercel Edge runtime because you've downcompiled it to `es2017`: https://github.com/solana-labs/solana-web3.js/issues/2595

The solution here is to compile `rpc-websockets` for the most modern target (`esnext`) and let your _consumers'_ bundlers downcompile to whatever target is appropriate for their audience.

### PR DESCRIPTION
**Pull request by steveluscher**
# Summary

First, read the rationale in #158.

This PR does the following:

1. Builds code for the highest target possible (`ESNext`) so as to give downstream consumers the _most_ choice of which target _they_ want to downcompile their application for.
2. Replaces Browserify with ESBuild
3. Builds _only_ the browser bundle (the iife one) with Babel (for `@babel/runtime` polyfills) targeting `ES2017`.

This PR will make downstream consumers' applications smaller, faster, and more compatible. For instance, this will make `rpc-websockets` compatible with Vercel Edge runtime, and will shrink bundle sizes for customers whose audience does not need `@babel/runtime` polyfills.

# Test Plan

```
npm run test
npm run test:client
npm run test:server
```

Closes #158.

### CODE REVIEW

**Comment by steveluscher** [comment]
cc/ @etodanik for additional input.

File: build-ts/lib/client.js [hunk 1]
```diff
@@ -254,7 +270,7 @@ export default class CommonClient extends EventEmitter {
                 if (message.method) {
                     // run as microtask so that pending queue messages are resolved first
                     return Promise.resolve().then(() => {
-                        this.emit(message.method, message === null || message === void 0 ? void 0 : message.params);
```
Comment (steveluscher) [inline build-ts/lib/client.js:257]: This is one of those classic examples of how building for a higher target can yield bundle size savings for a consumer that doesn't need es2017 compatibility.

**Comment by etodanik** [comment]
Agreed, this PR is sparked by trouble we had by utilizing an upstream lib that uses `rpc-websockets`, which brought babel runtime and cause havoc for us.

Most Edge runtimes (CloudFlare Workers, Vercel, Deno e.t.c) won't allow for dynamic code execution (which is what babel runtime does).

It'd be preferable indeed if the decision to do a lossy backwards compatibility transpilation was left to the downstream consumer.

**Comment by steveluscher** [comment]
This also results in an 83% reduction in the gzipped size of the browser bundle (the iife one).

**Comment by mkozjak** [comment]
Thanks for this work.

Is there something we need to adapt to make tests work?
`./build-browser-bundle.mjs` fails with `Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@babel/core' imported from /private/tmp/rpc-websockets/node_modules/esbuild-plugin-babel/src/index.js`.

**Comment by steveluscher** [comment]
~~Strange! I don't get that error. You definitely nuked `node_modules` and ran `npm i`, right?~~

EDIT: Oh, you said tests. Hold on, let me look.

**Comment by steveluscher** [comment]
Updated!

File: package.json [hunk 2]
```diff
@@ -27,20 +26,19 @@
     "utf-8-validate": "^5.0.2"
   },
   "devDependencies": {
-    "@babel/cli": "^7.17.0",
-    "@babel/core": "^7.17.2",
+    "@babel/core": "^7.24.5",
```
Comment (steveluscher) [inline package.json:29]: Turns out this was a `peerDependency` of `esbuild-plugin-babel`.

**Comment by steveluscher** [comment]
This just dropped a full 5% gzipped bytes from the iife bundle of `@solana/web3.js`. Thank you.

**Comment by etodanik** [comment]
Loving the rabbit hole I sent everyone to for the good of humanity

**Pull request merged**